homeassistant/components/heos/__init__.py

```diff
@@ -88,7 +103,100 @@
+        self.source_list = self._build_source_list()
+
+    def _build_source_list(self):
+        """Build a single list of inputs from various types."""
+        source_list = []
+        source_list.extend([favorite.name for favorite
+                            in self.favorites.values()])
+        source_list.extend([source.name for source in self.inputs])
+        return source_list
+
+    async def play_source(self, source: str, player):
+        """Determine type of source and play it."""
+        index = next((index for index, favorite in self.favorites.items()
+                      if favorite.name == source), None)
+        if index is not None:
+            return await player.play_favorite(index)
+
+        input_source = next((input_source for input_source in self.inputs
+                             if input_source.name == source), None)
+        if input_source is not None:
+            return await player.play_input_source(input_source)
+
+        raise ValueError("Unknown source: " + source)
+
+    def get_current_source(self, now_playing_media):
+        """Determine current source from now playing media."""
+        from pyheos import const
+        # Match input by input_name:media_id
+        if now_playing_media.source_id == const.MUSIC_SOURCE_AUX_INPUT:
+            return next((input_source.name for input_source in self.inputs
+                         if input_source.input_name ==
+                         now_playing_media.media_id), None)
+        # Try matching favorite by name:station or media_id:album_id
+        return next((source.name for source in self.favorites.values()
+                     if source.name == now_playing_media.station
+                     or source.media_id == now_playing_media.album_id), None)
+
+    def connect_update(self, hass, controller):
+        """
+        Connect listener for when sources change and signal player update.
+
+        EVENT_SOURCES_CHANGED is often raised multiple times in response to a
+        physical event therefore throttle it. Retrieving sources immediately
+        after the event may fail so retry.
+        """
+        from pyheos import CommandError, const
+
+        @Throttle(MIN_UPDATE_SOURCES)
+        async def get_sources():
+            retry_attempts = 0
+            while True:
+                try:
+                    return await asyncio.gather(
+                        controller.get_favorites(),
+                        controller.get_input_sources())
+                except (asyncio.TimeoutError, ConnectionError, CommandError):
+                    if retry_attempts < self.max_retry_attempts:
+                        retry_attempts += 1
+                        _LOGGER.debug("Error retrieving sources and will "
+                                      "retry", exc_info=True)
+                        await asyncio.sleep(self.retry_delay)
+                    else:
+                        _LOGGER.exception("Unable to update sources")
```

Comment:
Error level.

Reply:
Why don’t we want to log the trace back?  These APIs are orechstrstion of several lower level ones, which if it’s failing would be helpful to know where.  The lib won’t log this as it raised the exception.

Reply:
If we know the problem, we should log an error that informs the user of that. We don't need the stack trace.

Reply:
We know the impact (can't update sources), but not the lower level problem (why the controller failed and which command failed).  How would we get this information for troubleshooting otherwise?

Reply:
The controller should raise the error at sufficient granularity so that we know what the problem is. Otherwise raising the error isn't very useful.

Reply:
Agreed.  The lib could be improved with the introduction of a new exception type for these methods instead of relying on `CommandError `, whose message does include the controller error, but not all the data to reconstruct the command that failed w/out the trace back.

Would it be acceptable to move `CommandError ` to it's own `except` and log that with the trace?  Failing after 3 times should be _truly_ exceptional. I would then just log the message of `ConnectionError` and `Timeout`.

Reply:
Sounds good!

Reply:
Done.  I also tweaked the logging of during setup as well to use the same pattern.